go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+
+### ActivityInterceptor
+
+Proposal:
+
+```go
+type ActivityInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+}
+```
+
+Why: Mimics `WorkflowInterceptor` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
+}
+
+type ActivityInboundInterceptor interface {
+  Init(outbound ActivityOutboundInterceptor) error
+
+  ExecuteActivity(ctx context.Context, input *ActivityInput) (interface{}, error)
+}
+```
+
+Also a `ActivityInboundInterceptorBase` implementation with expected delegation.
+
+Why: Mimics `WorkflowInboundInterceptor` but for activities.
+
+### ActivityOutboundInterceptor
+
+Proposal:
+
+```go
+type ActivityOutboundInterceptor interface {
+  GetInfo(ctx context.Context) ActivityInfo
+  GetLogger(ctx Context) log.Logger
+  GetMetricsScope(ctx Context) tally.Scope
+  RecordHeartbeat(ctx context.Context, details ...interface{})
+  HasHeartbeatDetails(ctx context.Context) bool
+  GetHeartbeatDetails(ctx context.Context, d ...interface{}) error
+  GetWorkerStopChannel(ctx context.Context) <-chan struct{}
+}
+```
+
+Also a `ActivityOutboundInterceptorBase` implementation with expected delegation.
+
+Why:
+
+* Mimics `WorkflowOutboundInterceptor` but for activities.
+
+For discussion: `GetInfo` returns expensive copies because `activity.GetInfo` does. Is this ok? This is inconsistent
+with workflow info which is a pointer.
+
+## Interceptor Possible Updates
+
+### Forced Base Embedding
+
+Proposal: Force the base struct to be embedded in all interface implementations. This can be done via a private
+interface method that is only implemented by the base struct.
+
+Why: In the newer gRPC service interfaces this is done to allow interface additions to always be forward compatible. See
+[this](https://github.com/grpc/grpc-go/blob/03ca7b7d00cada2ff8a3ea7348fe6c3a2b2ee4fb/examples/helloworld/helloworld/helloworld_grpc.pb.go#L52)
+as an example.
+
+### WorkerInterceptor
+
+Current:
+
+```go
+type WorkflowInterceptor interface {
+  InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+}
+```
+
+Proposal:
+
+```go
+type WorkerInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+  InterceptWorkflow(ctx Context, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+}
+```
+
+Then change `WorkerOptions.WorkflowInterceptorChainFactories []WorkflowInterceptor` to
+`WorkerOptions.Interceptors []WorkerInterceptor`.
+
+For discussion:
+
+* This maps to what Java does
+* Would we want a `WorkerInterceptorBase` with default impls? Why doesn't Java have this?
+
+## OpenTelemetry Implementation
+
+Proposal in separate `opentelemetry` package:
+
+```go
+// Added to existing internal.ClientOptions struct
+type ClientOptions struct {
+  OpenTelemetry OpenTelemetryClientOptions
```

Comment:
Can you explain why this is needed please?

Reply:
To allow you to opt-in to opentelemetry gRPC and allow you to set current and future options implemented at https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc

Reply:
Explained in Godoc above the struct

Reply:
This tracing proposal has been removed and replaced after I learned more about how headers and tracing worked inside the SDK.